geopandas/clip.py

```diff
@@ -0,0 +1,246 @@
+    # the multi part object. Also reset index to remove the column made by explode.
+    clipped = _clip_line_poly(gdf.explode().reset_index(level=[1]), poly)
+
+    lines = clipped[
+        (clipped.geometry.type == "MultiLineString")
+        | (clipped.geometry.type == "LineString")
+    ]
+    line_diss = lines.dissolve(by=[lines.index]).drop(columns="level_1")
+
+    polys = clipped[clipped.geometry.type == "Polygon"]
+    poly_diss = polys.dissolve(by=[polys.index]).drop(columns="level_1")
+
+    return pd.concat([poly_diss, line_diss])
+
+
+def clip(gdf, clip_obj):
+    """Clip points, lines, or polygon geometries to the clip_obj extent.
+
+    Both layers must be in the same Coordinate Reference System (CRS) and will
+    be clipped to the full extent of the clip object.
+
+    If there are multiple polygons in clip_obj,
+    data from gdf will be clipped to the total boundary of
+    all polygons in clip_obj.
+
+    Parameters
+    ----------
+    gdf : GeoDataFrame
+          Vector layer (point, line, polygon) to be clipped to clip_obj.
+    clip_obj : GeoDataFrame, GeoSeries, (Multi)Polygon
+          Polygon vector layer used to clip gdf.
+          The clip_obj's geometry is dissolved into one geometric feature
+          and intersected with gdf.
+
+    Returns
+    -------
+    GeoDataFrame
+         Vector data (points, lines, polygons) from gdf clipped to
+         polygon boundary from clip_obj.
+
+    Examples
+    --------
+    Clip points (global capital cities) with a polygon (the South American continent):
+
+        >>> import geopandas as gpd
+        >>> import geopandas.clip as gc
+        >>> path = geopandas.datasets.get_path('naturalearth_lowres')
+        >>> world = geopandas.read_file(path)
+        >>> south_america = world[world['continent'] == "South America"]
+        >>> capitals = gpd.read_file(gpd.datasets.get_path('naturalearth_cities'))
+        >>> capitals.shape
+        (202, 2)
+        >>> sa_capitals = cl.clip(capitals, south_america)
+        >>> sa_capitals.shape
+        (12, 2)
+    """
+    if not isinstance(gdf, (gpd.GeoDataFrame, gpd.GeoSeries)):
+        raise TypeError(
+            "'gdf' should be GeoDataFrame or GeoSeries, got {}".format(type(gdf))
+        )
+
+    if not isinstance(
+        clip_obj, (gpd.GeoDataFrame, gpd.GeoSeries, Polygon, MultiPolygon)
+    ):
+        raise TypeError(
+            "'clip_obj' should be GeoDataFrame, GeoSeries or"
+            "(Multi)Polygon, got {}".format(type(gdf))
+        )
+
+    if isinstance(clip_obj, (gpd.GeoDataFrame, gpd.GeoSeries)):
+        xmin, ymin, xmax, ymax = clip_obj.total_bounds
+    else:
+        xmin, ymin, xmax, ymax = clip_obj.bounds
+    if gdf.cx[xmin:xmax, ymin:ymax].empty:
+        raise ValueError("Shape and crop extent do not overlap.")
+
+    poly = clip_obj.geometry.unary_union
+
+    geom_types = gdf.geometry.type
+    line_idx = np.asarray((geom_types == "Polygon") | (geom_types == "LineString"))
```

Comment:
```suggestion
    line_idx = np.asarray(
    (geom_types == "LineString")
    | (geom_types == "MultiLineString")
    | (geom_types == "LinearRing")
)
```

@lwasser This will cover the support for LinearRings, I've just tested it.

Reply:
@martinfleis cool! I'm working on implementing some of the changes that were discussed a few days ago right now, I'll be sure to throw this in.

Reply:
thank you @martinfleis !! much appreciated.